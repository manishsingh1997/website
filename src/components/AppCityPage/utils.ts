import {City} from './types';

export const getAsset = (src: string, ext: 'jpeg' | 'pdf') => {
  const filename = src.match(/file\/d\/([^/]+)/)?.[1];
  if (!filename) return src;
  return `https://s3.us-west-2.amazonaws.com/${process.env.S3_BUCKET}/cities_photos/${filename}.${ext}`;
};

export const makePhoneLink = (phone: string, options = {countryCode: '+1'}) => {
  let link = phone.replace(/[^\d+]/g, '');
  if (!/^\+/.test(link)) {
    link = `${options.countryCode}${link}`;
  }
  return `tel:${link}`;
};

export const formatFooterLicenses = (licenses?: City['licenses'], licenseUrl?: string) => {
  if (!licenses) return null;
  return licenses.map((license) => {
    return {
      name: `License #: ${license}`,
      url: licenseUrl,
    };
  });
};
