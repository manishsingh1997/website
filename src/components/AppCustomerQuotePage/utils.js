import {isUpcomingFeaturesEnabled} from '@ergeon/erg-utils-js';

export const QUOTE_LINE_STATUSES = {
  APPROVED: 'APPROVED',
  NEEDS_APPROVAL: 'NEEDS_APPROVAL',
  TO_BE_DROPPED: 'TO_BE_DROPPED',
};

export const prepareQuoteApprovalLines = (quoteApprovalLines, quote) => {
  return quoteApprovalLines.map((quoteApprovalLine) => {
    let status = QUOTE_LINE_STATUSES.NEEDS_APPROVAL;

    const quoteLine = quoteApprovalLine['quote_line'];

    const isDroppedBefore = quoteLine['is_dropped'] && quoteLine['dropped_at_quote_id'] !== quote.id;
    const isApprovedBefore = quoteLine['quote_id'] !== quote.id;
    const toBeDropped = quoteLine['is_dropped'] && quoteLine['dropped_at_quote_id'] === quote.id;

    if (toBeDropped) {
      status = QUOTE_LINE_STATUSES.TO_BE_DROPPED;
    } else if (isDroppedBefore || isApprovedBefore) {
      status = QUOTE_LINE_STATUSES.APPROVED;
    }

    return {
      ...quoteLine,
      percentage: quoteApprovalLine['percentage'],
      price: quoteApprovalLine['amount'],
      totalPrice: isUpcomingFeaturesEnabled('ENG-11541') ? quoteLine['display_price'] : quoteLine['price'],
      status,
    };
  });
};
