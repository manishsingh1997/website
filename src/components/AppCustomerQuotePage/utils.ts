import moment from 'moment';
import {Quote} from '../types';
import {DIRECT_PREVIEW_SLUG} from '../../website/constants';
import {updateCustomerSignOffRequirement} from '../../api/app';
import {QuoteApproval, SignatureData} from './types';

export const isSignOffPDFView = (pathname: string) => {
  return pathname.includes('/sign-off');
};

export const isDirectPreview = (path: string) => {
  return new RegExp(`${DIRECT_PREVIEW_SLUG}/?`).test(path);
};

export const isQuoteApprovalReviewed = (quote: Quote) => {
  return quote['reviewed_at'] !== null;
};

export const isQuoteApprovalApproved = (quote: Quote) => {
  return quote['approved_at'] !== null;
};

export const isScopeChange = (quoteApproval: QuoteApproval) => {
  return quoteApproval.quote && quoteApproval.quote['is_scope_change'] == true;
};

export const getTotalPrice = (quote: QuoteApproval) => {
  const totalPrice = Number.parseFloat(quote['total_price']);
  return totalPrice;
};

export const getProjectTotalPrice = (quoteApproval: QuoteApproval) => {
  return Number.parseFloat(quoteApproval['project_total_price']);
};

export const getTotalPreviouslyApprovedPrice = (quoteApproval: QuoteApproval) => {
  if (isScopeChange(quoteApproval)) {
    return Number.parseFloat(quoteApproval['previously_approved_price']);
  }
  return 0;
};

export const getSignatureData = async (
  customerGID: string,
  secret: string,
  value: string,
  type: string
): Promise<SignatureData | null> => {
  if (!value || !type) return null;

  const signOffPhrase = type === 'draw' ? 'signoff_img' : 'signoff_text';
  const {data} = await updateCustomerSignOffRequirement(customerGID, secret, {
    [signOffPhrase]: value,
    signoff_type: type,
  });
  const {signoff_img, signoff_at, signoff_pdf} = data || {};

  let newSignatureData = null;

  if (signoff_img) {
    newSignatureData = {
      value: signoff_img,
      type: 'draw', // so it can be displayed as image
      signedDate: signoff_at,
      signedPDF: signoff_pdf,
    };
  }

  return newSignatureData as SignatureData | null;
};

export const getIsQuoteStatusApprovedOrCompleted = (quoteApproval: QuoteApproval) => {
  const approvedAt = quoteApproval.quote.approved_at;
  const quoteStatus = quoteApproval.quote?.status?.label.toLocaleLowerCase();
  const quoteExpiredDate = quoteApproval.quote?.expires_at ||  moment().add('day', 2).toISOString();
  const isNotExpired = quoteExpiredDate
    ? moment(quoteExpiredDate).isAfter(new Date())
    : false;
  const allowedStatus = ['approved', 'completed'];
  return !!approvedAt && isNotExpired && allowedStatus.includes(quoteStatus);
};
