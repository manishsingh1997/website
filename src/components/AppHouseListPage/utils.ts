import {Address, HouseType} from '../types';

import {AddAddressProps} from './types';

export const parseAddressData = (data: {address: Address}): AddAddressProps | null => {
  try {
    return {
      raw_address: data.address.formatted_address,
      zip_code: data.address.zipcode,
      place_types: data.address.place_types,
      location: data.address.location,
    };
  } catch (error) {
    console.error(error);
    return null;
  }
};

export const getInitalAddress = (house: HouseType) => {
  try {
    const [address] = house.address.formatted_address.split(',');
    return address;
  } catch (error) {
    console.error(error);
    return null;
  }
};
