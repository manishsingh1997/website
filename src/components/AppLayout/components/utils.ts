/**
 * Function to extract a singular name + id from our pathName string.
 * Ex. /app/5QMHiaxMKUi--4wR/orders/12345 => Order #12345
 */
export const getSingularLevelFromPath = (pathname: string, pathToMatch: string) => {
  const matchRegex = new RegExp(`/${pathToMatch}/\\d+`); // we want to look for /orders/12345
  if (pathname.match(matchRegex)) {
    const [id, parentPath] = pathname.split('/').reverse(); // reversing so its easier to iterate 
    const singular = parentPath.slice(0, -1); // removes the 's'
    return `${singular} #${id}`;
  }
  return null;
}
