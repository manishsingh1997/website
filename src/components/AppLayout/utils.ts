import smartlookClient from 'smartlook-client';

export const initSmartLook = (smartlookID = process.env.SMARTLOOK_ID) => {
  if (process.env.NODE_ENV !== 'development' && smartlookID) {
    smartlookClient.init(smartlookID);
  }
};
