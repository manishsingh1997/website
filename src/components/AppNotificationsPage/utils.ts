import {Dispatch} from 'react';
import find from 'lodash/find';
import {updateNotificationPreferences} from '../../api/app';
import {Action, ActionType} from './appNotificationPageReducer';
import {COMPANY_NEWS_NOTIFICATION_TYPE} from './constant';
import {NotificationPreference} from './types';
import {updateNotificationPreferencesOnLocal} from './actions';

export const getNotificationPreferenceFromResponse = (
  response: {
    data: NotificationPreference[],
  }
): NotificationPreference | null => {
  const notificationPreference = find(response.data, 'is_primary');
  return notificationPreference ?? null;
}

/**
  * If `subscribeParameter` and `unsubscribeCode` are provided - it is a signal that user used "Unsubscribe" link
  * from marketing emails. In such case we should unsubscribe him from specific subscription.
*/
export const shouldUnsubscribe = (subscribeParameter: string | null, unsubscribeCode: string | null) => {
  return unsubscribeCode && subscribeParameter === COMPANY_NEWS_NOTIFICATION_TYPE;
}

export const fetchData = async (
    customerGID: string, 
    dispatch: Dispatch<Action>,
    notificationPreference: NotificationPreference | null,
    unsubscribeAutomatically: boolean
  ) => {
    await updateNotificationPreferencesOnLocal(customerGID, dispatch, location);
    await unsubscribeIfSubscribed(customerGID, dispatch, notificationPreference, unsubscribeAutomatically);
  }

export const unsubscribeIfSubscribed = async (
  customerGID: string, 
  dispatch: Dispatch<Action>,
  notificationPreference: NotificationPreference | null,
  unsubscribeAutomatically: boolean
) => {
  if (unsubscribeAutomatically) {
    let newState: NotificationPreference;
    if (notificationPreference) {
      dispatch({type: ActionType.UPDATE_IS_INITIAL_LOADING, isInitialLoading: true});
      newState = {...notificationPreference};
      newState.is_email_newsletter_ok = false;
      await updateNotificationPreferences(customerGID, dispatch, [newState]);
    }
  }
}
