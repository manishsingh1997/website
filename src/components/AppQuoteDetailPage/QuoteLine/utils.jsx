import React from 'react';
import PropTypes from 'prop-types';
import isEqual from 'lodash/isEqual';
import {getLabelFromIndex} from '@ergeon/draw-map';

import iconPhotoPlaceholder from '@ergeon/core-components/src/assets/icon-photo-placeholder.svg';

/**
 * Returns matching quoteLine comparing vs itemName
 * @param {Array.<object>} quoteLines
 * @param {string} itemName
 */
export const getQuoteLineForCalcInputItem = (quoteLines, itemName) => {
  if (!itemName) {
    return undefined;
  }
  return quoteLines.find((quoteLine) => Boolean(quoteLine.label) && isEqual(quoteLine.label, itemName));
};

/**
 * Returns label or retrieves one using index with getLabelFromIndex func
 * @param {number} index
 * @param {string} label
 */
export const indexLabel = (index, label) => {
  return label ? label : getLabelFromIndex(index);
};

/**
 * Checks for isInstallerPreview and catalog
 * @param {boolean} isInstallerPreview
 * @param {object} catalog
 */
export const isAllowedUnitDisplay = (isInstallerPreview, catalog) => {
  if (!isInstallerPreview || !catalog) {
    return true;
  }
  return catalog.type.allows_unit_display;
};

export const ProjectImagesLegend = ({imagesLen}) => {
  return (
    <div className="quote-project-images__legend">
      <img alt="" src={iconPhotoPlaceholder} />
      <span className="spacing left__is-5">
        Project Images {imagesLen > 0 && `(${imagesLen})`}
      </span>
    </div>
  );
};

ProjectImagesLegend.propTypes = {
  imagesLen: PropTypes.number,
};
