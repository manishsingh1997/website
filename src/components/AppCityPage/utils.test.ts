import cityData from '../../data/cities/alameda-county-berkeley-ca.json';
import {getAsset} from './utils';

describe('AppCityPage utils', () => {
  test('getAsset', () => {
    expect(
      getAsset(cityData.regulations.pdf, 'pdf')
    ).toMatchSnapshot();
    expect(
      getAsset(cityData.regulations.img, 'jpeg')
    ).toMatchSnapshot();
    expect(
      getAsset(cityData.regulations.url, 'pdf')
    ).toMatchSnapshot();
  });
});
