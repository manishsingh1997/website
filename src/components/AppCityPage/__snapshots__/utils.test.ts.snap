// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppCityPage utils getAsset 1`] = `"https://s3.us-west-2.amazonaws.com/dev.ergeon.com/cities_photos/1A_x5vu8J7uWmRycAfOqq5H7MzGRM3t4J.pdf"`;

exports[`AppCityPage utils getAsset 2`] = `"https://s3.us-west-2.amazonaws.com/dev.ergeon.com/cities_photos/1L-EulsROIe29OpNv-R5pzeQz3sq6IheS.jpeg"`;

exports[`AppCityPage utils getAsset 3`] = `"https://www.codepublishing.com/CA/Berkeley/html/Berkeley23D/Berkeley23D08/Berkeley23D08060.html"`;
